### PR TITLE
Bring life cycle in line with npm standards

### PR DESCRIPTION
**Pull request by johnhaley81**
We were doing our own highly custom (read: fragile) life cycle scripting for installing NodeGit.

This should make it more in line with npm standards.

### CODE REVIEW

**Comment by johnhaley81** [comment]
@tbranyen @maxkorp we're passing CI. Can you guys give some 👀 on this?

**Comment by tbranyen** [comment]
I'll test it tonight :heart_eyes:

**Comment by maxkorp** [comment]
Talked with john about this in person a moment ago, but putting here for records. 

We can't clean the source afterwards.

The standard workflow for electron/nwjs is to npm install normally, then run a rebuild with the appropriate headers. This works with all other native modules i've seen, and currently works with nodegit, but cleaning the src/includes out would break that. We have to be able to run `npm run rebuild` after install has finished. 

Otherwise, gonna review it up right now.

File: .npmignore [hunk 1]
```diff
@@ -1,11 +1,14 @@
+/.travis/
 /build/
 /examples/
+/generate/
+/guides/
+/lib/
```
Comment (maxkorp) [inline .npmignore]: Why ignore lib? all of our js is in there?

File: generate/scripts/generateJson.js [hunk 2]
```diff
@@ -222,7 +222,7 @@ module.exports = function generateJson() {
     _.merge(enumerable, _.omit(override, ["values"]));
 
     output.push(enumerable);
-  }).value();
+  });
```
Comment (maxkorp) [inline generate/scripts/generateJson.js]: don't we need the `.value` to make lodash actually do the operation? I thought they were lazy about that as of v4?

File: .npmignore (hunk 1 above)
Comment (johnhaley81) [inline .npmignore]: lib should already be 100% built and in dist.

File: generate/scripts/generateJson.js (hunk 2 above)
Comment (johnhaley81) [inline generate/scripts/generateJson.js]: Correct. We needed it in v3 but in v4 a `forEach` just executes.

Comment (maxkorp) [inline generate/scripts/generateJson.js]: PRAISE BE TO LODASH, FOR THEY ARE IMPROVE

File: .npmignore (hunk 1 above)
Comment (maxkorp) [inline .npmignore]: That makes complete and total sense and I should have realized that, but sometimes I am do the forget.

File: lifecycleScripts/postinstall.js [hunk 3]
```diff
@@ -20,13 +20,33 @@ function printStandardLibError() {
...
+      }
+    })
     .then(function() {
       // Is we're using NodeGit from a package manager then let's clean up after
       // ourselves when we install successfully.
       if (!buildFlags.mustBuild) {
-        fse.removeSync(path.join(rootPath, "vendor"));
-        fse.removeSync(path.join(rootPath, "src"));
-        fse.removeSync(path.join(rootPath, "include"));
+        // We can't remove the source files yet because apparently the
```
Comment (maxkorp) [inline lifecycleScripts/postinstall.js]: This is insufferably true :(

**Comment by maxkorp** [comment]
Good change, better comment.

**Comment by maxkorp** [comment]
This should be good now :)

File: package.json [hunk 4]
```diff
@@ -51,11 +52,7 @@
     "js-beautify": "~1.5.10",
     "jshint": "~2.8.0",
     "lcov-result-merger": "~1.0.2",
-    "lodash": "~3.10.1",
-    "mocha": "~2.3.4",
-    "nan": "^2.2.0",
-    "node-gyp": "~3.0.3",
```
Comment (maxkorp) [inline package.json]: The gyp exclusions make mme SO GODDAMNED HAPPY.

**Comment by tbranyen** [comment]
I'm so far removed from the lifecycle hooks code lately, but this looks v. good, nice work!

**Comment by maxkorp** [comment]
@tbranyen There are a lot of improvements in npm3, but there are some definite regressions, and it doesn't fix the fact that npm2 is bundled with node 4 which is LTS :/

By excluding having to generate the code in our bundles, we just have to configure libssh2 if needed, ensure our deps are all installed when we run node-pre-gyp (this is not guaranteed properly in npm2, but it is in npm3 and trying to insure it ourself anyways actually breaks in npm3). We fall back to the default node-gyp now, since all that confusion was from A) node 0.8 and then came back in iojs 1/2. 

A lot has changed, but overall the process is easier to follow now in npm3.

**Comment by tbranyen** [comment]
:tada:

**Pull request merged**